src/Plutarch/Test/QuickCheck.hs

```diff
@@ -213,60 +326,120 @@ type family PLamArgs (p :: S -> Type) :: [Type] where
+            Right (Left err, _, t) ->
+                case h of
+                    Just _ -> failWith $ "Haskell expected success, but Plutarch evaluation failed.\n" <> show err <> "\n" <> show t
+                    Nothing -> property True
+            Right (Right p', _, t) ->
+                case h of
+                    Just h' -> haskEquiv @e @( 'ByComplete) h' (TestableTerm p') Nil
+                    Nothing -> failWith $ "Haskell expected failure, but Plutarch succeed.\n" <> show t
+
+-- | @since 2.1.0
+instance
+    forall (p :: S -> Type) (h :: Type).
+    (PLamArgs p ~ '[], PLift p, PLifted p ~ h, PEq p) =>
+    HaskEquiv 'OnPEq 'ByComplete h p '[]
+    where
+    haskEquiv h (TestableTerm p) _ =
+        counterexample "Comparison by PEq Failed" $
+            property $ plift $ p #== (pconstant h)
+
+-- | @since 2.1.0
 instance
     forall (p :: S -> Type) (h :: Type).
-    (PLamArgs p ~ '[], PLift p, PLifted p ~ h, Eq h) =>
-    HaskEquiv h p '[]
+    (PLamArgs p ~ '[], PLift p, PLifted p ~ h, PIsData p) =>
+    HaskEquiv 'OnPData 'ByComplete h p '[]
     where
-    haskEquiv h (TestableTerm p) _ = property $ plift p == h
+    haskEquiv h (TestableTerm p) _ =
+        counterexample "Comparison by PData Failed" $
+            property $ plift (pdata p #== pdata (pconstant h))
+
+-- | @since 2.1.0
+instance
+    forall (p :: S -> Type) (h :: Type).
+    (PLamArgs p ~ '[], PLift p, PLifted p ~ h, PIsData p, PEq p) =>
+    HaskEquiv 'OnBoth 'ByComplete h p '[]
+    where
+    haskEquiv h p _ =
+        (haskEquiv @( 'OnPEq) @( 'ByComplete) h p Nil) .&&. (haskEquiv @( 'OnPData) @( 'ByComplete) h p Nil)
 
 {- | Simplified version of `haskEquiv`. It will use arbitrary instead of
      asking custom generators.
 
  @since 2.0.0
 -}
 haskEquiv' ::
-    forall (h :: Type) (p :: S -> Type) (args :: [Type]).
+    forall (e :: Equality) (par :: Partiality) (h :: Type) (p :: S -> Type) (args :: [Type]).
     ( PLamArgs p ~ args
-    , HaskEquiv h p args
+    , HaskEquiv e par h p args
     , All Arbitrary args
     ) =>
     h ->
     (forall s. Term s p) ->
     Property
 haskEquiv' h p =
-    haskEquiv h (TestableTerm p) $
-        hcpure
-            (Proxy @Arbitrary)
-            arbitrary
+    haskEquiv @e @par h (TestableTerm p) $
+        hcpure (Proxy @Arbitrary) arbitrary
+
+{- | Compares evaluated UPLC
+
+ @since 2.0.1
+-}
+uplcEq :: TestableTerm a -> TestableTerm b -> Property
+uplcEq x y = either failWith property go
+  where
+    go = do
+        x' <- eval x
+        y' <- eval y
+        return $ x' == y'
+    eval (TestableTerm t) =
+        case compile (Config{tracingMode = NoTracing}) t of
+            Left err -> Left $ "Term compilation failed:\n" <> show err
+            Right s' ->
+                case evalScript s' of
+                    (Right s, _, _) -> Right s
+                    (Left err, _, _) -> Left $ "Term evaluation failed:\n" <> show err
 
 newtype A = A Integer
```

Comment:
Can't we re-use QuickCheck's one?

Reply:
Because we need to evaluate twice, this gets very messy if you do that.

Reply:
We should probably publically expose this, along with `B` and `C`, and encourage people to use those instead of QuickCheck's ones.

Reply:
Oh sorry, I thought you are talking about `uplcEq`. We can make use of QuickCheck `A`. Sorry about the confusion.